File: parity/streamlit/go/apps.go
package main

import (
	"github.com/malcolmston/streamlit/st"
)

// appCtx is the side channel through which a Go app publishes what it observed.
//
// Streamlit's oracle reads the app's own session-state keys straight out of
// st.session_state; the Go port's published API (v0.3.0) has no way to
// enumerate a [st.State] — no Keys, Len or iteration — so each app names the
// keys it publishes explicitly. See COVERAGE.md.
type appCtx struct {
	state map[string]any
	// computed counts real (uncached) computations for the caching apps. It
	// lives on the context rather than in a package variable so each case is
	// independent.
	computed int
}

func newAppCtx() *appCtx { return &appCtx{state: map[string]any{}} }

// put stores a value in the session's state and publishes it to the runner.
func (a *appCtx) put(s *st.Session, key string, v any) {
	s.State.Set(key, v)
	a.state[key] = v
}

// pub publishes a value to the runner without touching session state. It is
// used for values the app merely observed rather than persisted.
func (a *appCtx) pub(key string, v any) { a.state[key] = v }

// widgetSnippets holds one widget call per entry, mirroring
// python/apps.py:WIDGET_SNIPPETS. Composing them keeps the single-widget apps
// and the combined app in step on both sides.
var widgetSnippets = map[string]func(*appCtx, *st.Session){
	"button": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.Button("Go", "w_button"))
	},
	"checkbox": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.Checkbox("Agree", true, "w_checkbox"))
	},
	"radio": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.Radio("Pick", []string{"alpha", "beta", "gamma"}, "w_radio"))
	},
	"selectbox": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.SelectBox("Choose", []string{"alpha", "beta", "gamma"}, "w_selectbox"))
	},
	"multiselect": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", sortedStrings(s.MultiSelect("Many", []string{"alpha", "beta", "gamma"}, "w_multiselect")))
	},
	"slider": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.Slider("Amount", 0, 10, 3, 1, "w_slider"))
	},
	"text_input": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.TextInput("Name", "ada", "w_text_input"))
	},
	"text_area": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.TextArea("Bio", "hello", "w_text_area"))
	},
	"number_input": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.NumberInput("Count", 5, "w_number_input"))
	},
	"date_input": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.DateInput("Day", "2020-01-02", "w_date_input"))
	},
	"time_input": func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.TimeInput("Moment", "09:30", "w_time_input"))
	},
	"file_uploader": func(a *appCtx, s *st.Session) {
		files := s.FileUploader("Upload", "w_file_uploader")
		names := make([]string, 0, len(files))
		for _, f := range files {
			names = append(names, f.Name)
		}
		a.put(s, "ss_v", names)
	},
}

// widgetOrder fixes the order the combined widgets-defaults app draws its
// widgets in, matching the insertion order of python/apps.py:WIDGET_SNIPPETS.
var widgetOrder = []string{
	"button", "checkbox", "radio", "selectbox", "multiselect", "slider",
	"text_input", "text_area", "number_input", "date_input", "time_input",
	"file_uploader",
}

func sortedStrings(xs []string) []string {
	out := make([]string, 0, len(xs))
	out = append(out, xs...)
	for i := 1; i < len(out); i++ {
		for j := i; j > 0 && out[j] < out[j-1]; j-- {
			out[j], out[j-1] = out[j-1], out[j]
		}
	}
	return out
}

// apps is the Go half of the logical app registry. Names must match
// python/apps.py:APPS exactly.
var apps = map[string]func(*appCtx, *st.Session){}

func init() {
	for name, fn := range widgetSnippets {
		apps["wd-"+name] = fn
	}
	apps["widgets-defaults"] = func(a *appCtx, s *st.Session) {
		for _, name := range widgetOrder {
			widgetSnippets[name](a, s)
		}
	}

	apps["text-elements"] = func(a *appCtx, s *st.Session) {
		s.Title("Parity")
		s.Header("Section")
		s.Subheader("Subsection")
		s.Markdown("**bold** and `code`")
		s.Caption("a caption")
		s.Latex("x^2 + y^2")
		s.Code("print(1)", "python")
		s.Text("plain text")
		s.Divider()
		s.Success("all good")
		s.Info("for your information")
		s.Warning("be careful")
		s.Error("it broke")
	}

	apps["write-dispatch"] = func(a *appCtx, s *st.Session) {
		s.Write("a string")
		s.Write(3)
		s.Write(true)
	}

	// The Go port has no help, disabled, placeholder or format parameter on any
	// widget; TextInputMax is the only one of the five concepts it models.
	apps["widgets-labelled"] = func(a *appCtx, s *st.Session) {
		s.Checkbox("Agree", true, "w_checkbox")
		s.Slider("Amount", 0, 10, 3, 1, "w_slider")
		s.TextInputMax("Name", "ada", 8, "w_text_input")
	}

	apps["slider-bounds"] = func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.Slider("Amount", 0, 10, 3, 1, "w_slider"))
	}
	apps["number-bounds"] = func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.NumberInputRange("Count", 0, 10, 5, "w_number_input"))
	}
	apps["selectbox-unknown"] = func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", s.SelectBox("Choose", []string{"alpha", "beta", "gamma"}, "w_selectbox"))
	}
	apps["multiselect-unknown"] = func(a *appCtx, s *st.Session) {
		a.put(s, "ss_v", sortedStrings(s.MultiSelect("Many", []string{"alpha", "beta", "gamma"}, "w_multiselect")))
	}

	apps["layout"] = func(a *appCtx, s *st.Session) {
		cols := s.Columns(2)
		cols[0].Text("left")
		cols[1].Text("right")

		tabs := s.Tabs([]string{"One", "Two"})
		tabs[0].Text("in one")
		tabs[1].Text("in two")

		s.Expander("More", true).Text("inside expander")
		s.Container().Text("inside container")
		s.Sidebar().Text("in sidebar")
	}

	apps["layout-nested"] = func(a *appCtx, s *st.Session) {
		cols := s.Columns(2)
		cols[0].Expander("Nested", false).Text("deep")
		cols[1].Text("shallow")
	}

	apps["layout-widget-in-column"] = func(a *appCtx, s *st.Session) {
		cols := s.Columns(2)
		a.put(s, "ss_v", cols[0].TextInput("Name", "ada", "w_text_input"))
		cols[1].Text("right")
	}

	// The Go port takes tabular data as [][]string with the header in row 0;
	// Streamlit takes a list of records. Same logical table either way.
	rows := [][]string{{"name", "qty"}, {"ada", "1"}, {"bob", "2"}}
	apps["dataframe"] = func(a *appCtx, s *st.Session) { s.DataFrame(rows) }
	apps["table"] = func(a *appCtx, s *st.Session) { s.Table(rows) }
	apps["json-element"] = func(a *appCtx, s *st.Session) {
		s.JSON(map[string]any{"b": 2, "a": []int{1, 2}})
	}

	apps["metric"] = func(a *appCtx, s *st.Session) {
		s.Metric("Revenue", "1.2M", "+4.1%")
		s.Metric("Latency", "120ms", "-8ms")
		s.Metric("Flat", "0", "0")
	}

	apps["chart-line"] = func(a *appCtx, s *st.Session) { s.LineChart([]float64{1, 3, 2}) }
	apps["chart-area"] = func(a *appCtx, s *st.Session) { s.AreaChart([]float64{1, 3, 2}) }
	apps["chart-bar"] = func(a *appCtx, s *st.Session) { s.BarChart([]float64{1, 3, 2}) }
	apps["chart-scatter"] = func(a *appCtx, s *st.Session) {
		s.ScatterChart([]float64{1, 2, 3}, []float64{4, 5, 6})
	}

	apps["form-atomic"] = func(a *appCtx, s *st.Session) {
		f := s.Form("signup")
		name := f.TextInput("Name", "", "f_name")
		age := f.NumberInput("Age", 18, "f_age")
		submitted := f.FormSubmitButton("Register")
		a.put(s, "ss_name", name)
		a.put(s, "ss_age", age)
		a.put(s, "ss_submitted", submitted)
	}

	apps["session-state"] = func(a *appCtx, s *st.Session) {
		a.put(s, "ss_runs", s.State.GetInt("ss_runs", 0)+1)
		clicks := s.State.GetInt("ss_clicks", 0)
		if s.Button("Bump", "w_button") {
			clicks++
		}
		a.put(s, "ss_clicks", clicks)
	}

	apps["button-one-run"] = func(a *appCtx, s *st.Session) {
		a.put(s, "ss_hit", s.Button("Fire", "w_button"))
		a.put(s, "ss_runs", s.State.GetInt("ss_runs", 0)+1)
	}

	apps["widget-hidden-then-shown"] = func(a *appCtx, s *st.Session) {
		if s.Checkbox("Show", true, "w_flag") {
			a.put(s, "ss_v", s.TextInput("Name", "ada", "w_text_input"))
		} else {
			a.put(s, "ss_v", nil)
		}
	}

	apps["cache-hits"] = func(a *appCtx, s *st.Session) {
		n := s.NumberInput("N", 2, "w_number_input")
		out := s.Cache(cacheKey("double", n), func() any {
			a.computed++
			return n * 2
		})
		a.pub("ss_computed", a.computed)
		a.pub("ss_out", out)
	}

	apps["cache-invalidate"] = func(a *appCtx, s *st.Session) {
		if s.Button("Clear", "w_button") {
			st.CacheClear()
		}
		out := s.Cache("constant", func() any {
			a.computed++
			return a.computed
		})
		a.pub("ss_computed", a.computed)
		a.pub("ss_out", out)
	}

	apps["stop"] = func(a *appCtx, s *st.Session) {
		s.Text("before")
		if s.Checkbox("Halt", false, "w_checkbox") {
			s.Stop()
		}
		s.Text("after")
	}

	// Session.Stop halts by panicking an unexported sentinel, so a recover() in
	// app code changes control flow. This app is the analogue of wrapping
	// st.stop() in `try: ... except BaseException:`.
	apps["stop-swallowed"] = func(a *appCtx, s *st.Session) {
		s.Text("before")
		func() {
			defer func() {
				if r := recover(); r != nil {
					s.Text("swallowed")
				}
			}()
			s.Stop()
		}()
		s.Text("after")
	}

	apps["auto-keys"] = func(a *appCtx, s *st.Session) {
		if s.Checkbox("Show extra", false, "w_flag") {
			s.Slider("Extra", 0, 10, 0, 1)
		}
		a.put(s, "ss_target", s.Slider("Target", 0, 100, 5, 1))
	}

}

// cacheKey builds the explicit cache key the Go port requires. Streamlit's
// @st.cache_data hashes the arguments itself; [st.Session.Cache] takes a
// caller-derived key, so the argument has to be spliced in by hand.
func cacheKey(name string, n float64) string {
	return name + ":" + trimFloat(n)
}

File: parity/streamlit/python/apps.py
"""The upstream half of every logical app the parity harness compares.

Each entry is a Streamlit script, executed headlessly through
``streamlit.testing.v1.AppTest.from_string``. The Go half of the same logical
app lives in ``../go/apps.go`` and must produce the same element tree.

Conventions shared with the Go side:

* Every widget a case needs to drive carries an explicit ``key``, except in the
  ``auto-keys`` app, which exists precisely to probe key generation.
* What the app *observed* (the value a widget call returned) is published into
  session state under an ``ss_`` prefix rather than rendered as text, so the
  two runners compare JSON numbers instead of two languages' float formatting.
"""

# --------------------------------------------------------------- widget snippets
#
# One line per widget, so the same source can be assembled into a
# single-widget app (granular scoring) and into the combined app.

WIDGET_SNIPPETS: dict[str, str] = {
    "button": 'st.session_state["ss_v"] = bool(st.button("Go", key="w_button"))',
    "checkbox": 'st.session_state["ss_v"] = st.checkbox("Agree", value=True, key="w_checkbox")',
    "radio": 'st.session_state["ss_v"] = st.radio("Pick", ["alpha", "beta", "gamma"], key="w_radio")',
    "selectbox": 'st.session_state["ss_v"] = st.selectbox("Choose", ["alpha", "beta", "gamma"], key="w_selectbox")',
    "multiselect": 'st.session_state["ss_v"] = sorted(st.multiselect("Many", ["alpha", "beta", "gamma"], key="w_multiselect"))',
    "slider": 'st.session_state["ss_v"] = st.slider("Amount", min_value=0.0, max_value=10.0, value=3.0, step=1.0, key="w_slider")',
    "text_input": 'st.session_state["ss_v"] = st.text_input("Name", value="ada", key="w_text_input")',
    "text_area": 'st.session_state["ss_v"] = st.text_area("Bio", value="hello", key="w_text_area")',
    "number_input": 'st.session_state["ss_v"] = st.number_input("Count", value=5.0, key="w_number_input")',
    "date_input": 'st.session_state["ss_v"] = st.date_input("Day", value=datetime.date(2020, 1, 2), key="w_date_input").isoformat()',
    "time_input": 'st.session_state["ss_v"] = st.time_input("Moment", value=datetime.time(9, 30), key="w_time_input").strftime("%H:%M")',
    "file_uploader": (
        '_f = st.file_uploader("Upload", key="w_file_uploader")\n'
        'st.session_state["ss_v"] = [] if _f is None else [_f.name]'
    ),
}

PREAMBLE = "import streamlit as st\nimport datetime\n\n"

APPS: dict[str, str] = {}

for _name, _src in WIDGET_SNIPPETS.items():
    APPS["wd-" + _name] = PREAMBLE + _src + "\n"

APPS["widgets-defaults"] = PREAMBLE + "\n".join(WIDGET_SNIPPETS.values()) + "\n"


def _add(name: str, body: str) -> None:
    APPS[name] = PREAMBLE + body.strip("\n") + "\n"


# --------------------------------------------------------------- text elements

_add(
    "text-elements",
    """
st.title("Parity")
st.header("Section")
st.subheader("Subsection")
st.markdown("**bold** and `code`")
st.caption("a caption")
st.latex("x^2 + y^2")
st.code("print(1)", language="python")
st.text("plain text")
st.divider()
st.success("all good")
st.info("for your information")
st.warning("be careful")
st.error("it broke")
""",
)

_add(
    "write-dispatch",
    """
st.write("a string")
st.write(3)
st.write(True)
""",
)

# --------------------------------------------------- widget presentation props

_add(
    "widgets-labelled",
    """
st.checkbox("Agree", value=True, key="w_checkbox", help="tick it", disabled=True)
st.slider(
    "Amount", min_value=0.0, max_value=10.0, value=3.0, step=1.0, format="%.1f",
    key="w_slider", help="slide it",
)
st.text_input(
    "Name", value="ada", key="w_text_input", max_chars=8, placeholder="who?",
    help="type it", disabled=True,
)
""",
)

# ------------------------------------------------------- clamping / validation

_add(
    "slider-bounds",
    """
st.session_state["ss_v"] = st.slider(
    "Amount", min_value=0.0, max_value=10.0, value=3.0, step=1.0, key="w_slider"
)
""",
)

_add(
    "number-bounds",
    """
st.session_state["ss_v"] = st.number_input(
    "Count", min_value=0.0, max_value=10.0, value=5.0, key="w_number_input"
)
""",
)

_add(
    "selectbox-unknown",
    """
st.session_state["ss_v"] = st.selectbox(
    "Choose", ["alpha", "beta", "gamma"], key="w_selectbox"
)
""",
)

_add(
    "multiselect-unknown",
    """
st.session_state["ss_v"] = sorted(
    st.multiselect("Many", ["alpha", "beta", "gamma"], key="w_multiselect")
)
""",
)

# ----------------------------------------------------------------------- layout

_add(
    "layout",
    """
left, right = st.columns(2)
left.text("left")
right.text("right")

first, second = st.tabs(["One", "Two"])
first.text("in one")
second.text("in two")

with st.expander("More", expanded=True):
    st.text("inside expander")

with st.container():
    st.text("inside container")

st.sidebar.text("in sidebar")
""",
)

_add(
    "layout-nested",
    """
cols = st.columns(2)
with cols[0]:
    with st.expander("Nested", expanded=False):
        st.text("deep")
cols[1].text("shallow")
""",
)

_add(
    "layout-widget-in-column",
    """
cols = st.columns(2)
st.session_state["ss_v"] = cols[0].text_input("Name", value="ada", key="w_text_input")
cols[1].text("right")
""",
)

# ------------------------------------------------------------- tabular and json

_add(
    "dataframe",
    """
st.dataframe([{"name": "ada", "qty": "1"}, {"name": "bob", "qty": "2"}])
""",
)

_add(
    "table",
    """
st.table([{"name": "ada", "qty": "1"}, {"name": "bob", "qty": "2"}])
""",
)

_add(
    "json-element",
    """
st.json({"b": 2, "a": [1, 2]})
""",
)

_add(
    "metric",
    """
st.metric("Revenue", "1.2M", "+4.1%")
st.metric("Latency", "120ms", "-8ms")
st.metric("Flat", "0", "0")
""",
)

# ----------------------------------------------------------------------- charts

_add("chart-line", 'st.line_chart([1.0, 3.0, 2.0])')
_add("chart-area", 'st.area_chart([1.0, 3.0, 2.0])')
_add("chart-bar", 'st.bar_chart([1.0, 3.0, 2.0])')
_add(
    "chart-scatter",
    'st.scatter_chart({"x": [1.0, 2.0, 3.0], "y": [4.0, 5.0, 6.0]}, x="x", y="y")',
)

# ------------------------------------------------------------------------ forms

_add(
    "form-atomic",
    """
with st.form("signup"):
    name = st.text_input("Name", value="", key="f_name")
    age = st.number_input("Age", value=18.0, key="f_age")
    submitted = st.form_submit_button("Register")
st.session_state["ss_name"] = name
st.session_state["ss_age"] = age
st.session_state["ss_submitted"] = bool(submitted)
""",
)

# ---------------------------------------------------------------- session state

_add(
    "session-state",
    """
st.session_state["ss_runs"] = st.session_state.get("ss_runs", 0) + 1
if "ss_clicks" not in st.session_state:
    st.session_state["ss_clicks"] = 0
if st.button("Bump", key="w_button"):
    st.session_state["ss_clicks"] += 1
""",
)

_add(
    "button-one-run",
    """
st.session_state["ss_hit"] = bool(st.button("Fire", key="w_button"))
st.session_state["ss_runs"] = st.session_state.get("ss_runs", 0) + 1
""",
)

_add(
    "widget-hidden-then-shown",
    """
show = st.checkbox("Show", value=True, key="w_flag")
if show:
    st.session_state["ss_v"] = st.text_input("Name", value="ada", key="w_text_input")
else:
    st.session_state["ss_v"] = None
""",
)

# ---------------------------------------------------------------------- caching

_add(
    "cache-hits",
    """
if "ss_computed" not in st.session_state:
    st.session_state["ss_computed"] = 0


@st.cache_data
def double(n):
    st.session_state["ss_computed"] += 1
    return n * 2


n = st.number_input("N", value=2.0, key="w_number_input")
st.session_state["ss_out"] = double(round(n, 3))
""",
)

_add(
    "cache-invalidate",
    """
if "ss_computed" not in st.session_state:
    st.session_state["ss_computed"] = 0


@st.cache_data
def constant():
    st.session_state["ss_computed"] += 1
    return st.session_state["ss_computed"]


if st.button("Clear", key="w_button"):
    constant.clear()
st.session_state["ss_out"] = constant()
""",
)

# --------------------------------------------------------- stop / control flow

_add(
    "stop",
    """
st.text("before")
if st.checkbox("Halt", value=False, key="w_checkbox"):
    st.stop()
st.text("after")
""",
)

_add(
    "stop-swallowed",
    """
st.text("before")
try:
    st.stop()
except BaseException:
    st.text("swallowed")
st.text("after")
""",
)

# ------------------------------------------------------------------- auto keys

_add(
    "auto-keys",
    """
if st.checkbox("Show extra", value=False, key="w_flag"):
    st.slider("Extra", min_value=0.0, max_value=10.0, value=0.0, step=1.0)
st.session_state["ss_target"] = st.slider(
    "Target", min_value=0.0, max_value=100.0, value=5.0, step=1.0
)
""",
)
